### ISSUE TITLE
Mixed-type fields?

### ISSUE THREAD

**Issue by CookieJon** [vuex-orm/vuex-orm#50]
Is it possible to define a field on a Model that may contain more than one type of Model, or an array containing different types of Models.

Something like normalizr's schema mapping function, giving control over how entities are contructed/deconstructed: 
https://github.com/paularmstrong/normalizr/blob/master/docs/api.md#schema  
 "If your input data is an array of more than one type of entity, it is necessary to define a schema mapping."

Thanks.

**Comment by kiaking** [comment]
It sounds interesting. I would love to look into this more. However, I'm not sure of your desired use case.

Could you give me some example of how it might look like to use this feature? Like how to define Model and what kind of output is expected through getters or actions?

**Comment by CookieJon** [comment]
Sure. Thanks for responding. Let me get my thoughts together and get back to you.

By the way,  thanks for this library! I was struggling this stuff for ages... wasted huge amounts of time trying to invent a solution myself, before realising it had already been solved. Tried "npormalizr" but it didn't go far enough... this ORM+VUEX combination is the first library I've actually been able to make inroads with.

**Comment by kiaking** [comment]
Oh no, thank "you"! for using this library and giving me feedback! Those kind words are the best reward I can get :)

Looking forward to seeing your example!

**Comment by CookieJon** [comment]
Hi Again,

Okay, I've had a think and I'm back. :-)   

Couldn't work out what I'm really trying to ask most, so I've put my use-case down in one big example with a few "how-to" questions along the way.  Now I realise this is a lot, so if it's not appropriate to ask for this kind of help/suggestion in this form I completely understand.

Anyway, here's a simplified & contrived example of what I'm trying to do...

1.  MODELS
```
  Bitmap {
    id: null,
    name: null,
    pixels: null,
    palette: Palette,
    filters: [PixelMap|ColorMap...]
  }
  Palette: {
    id: null,
    name: null,
    colors: [Color...]
  }
  Color: {
    id: null,
    name: null, 
    r: null,     g: null,     b: null  
  }
  PixelMap: {
    id: null,
    name: null,
    matrix: null
  }
  ColorMap: {
    id: null,
    name: null,
    colors: [Color...]
  }

Notes: 
1. Bitmaps have Palettes, and different Bitmaps may share the same Palette.
2. Palettes & ColorMaps have Colors & may share the same Color.
3. 'Bitmap.filters' is an array that can contain a mixture of PixelMaps AND ColorMaps.
```

2. DATA/ENTITIES
```
entities: {

  bitmaps: {
    1: {id: 1, name: 'Mona Lisa', palette: 0, pixels: [...], filters: [{type: PixelMap, id: 0}, {type: ColorMap, id: 1}] },
    2: {id: 2, name: 'Scream', palette: 1, pixels: [...], filters: [{type: PixelMap, id: 0}] },
    3: {id: 3, name: 'Nude Descending a Staircase', palette: 1,  pixels: [...], filters: [{type: ColorMap, id: 1}] }
  },

  palettes: {
    0: { id: 0, name: 'Greyscale', colors: [ 3, 4, 5 ] },
    1: { id: 1, name: 'Colorful', colors: [ 0, 1, 2, 3, 4 ] },
  },

  colors: {
    0: { id: 0, name: 'Red',    r: 255, g: 0, b: 0 },
    1: { id: 1, name: 'Orange', r: 255, g: 110, b: 0 },
    2: { id: 2, name: 'Yellow', r: 200, g: 200, b: 60 },
    3: { id: 3, name: 'White',  r: 255, g: 255, b:255 },
    4: { id: 4, name: 'Black',  r: 0, g: 0, b: 0 },
    5: { id: 5, name: 'Grey',   r: 122, g: 122, b: 122 }
  },

  pixelMaps: {
    0: { id: 0, type: 'pixelmap', name: 'Blur', matrix: [0,1,1,0,0,1,0,...] },
    1: { id: 1, type: 'pixelmap', name: 'Sharpen', matrix: [1,0,0,1,1,0,1...] }
  },

  colorMaps: {
    0: { id: 0, type: 'colormap', name: 'Warmer', colors: [0, 1] },
    1: { id: 1, type: 'colormap', name: 'Bichromal', colors: [3, 5] },
  }

}
```

3. "Hey, ORM! Fetch me the Mona Lisa!" 
```
RESULT: {
  id: 1, 
  name: 'Mona Lisa', 
  palette: { 
    id: 0, 
    name: 'Greyscale', 
    colors: [ 
      { id: 3, name: 'White',  r: 255, g: 255, b:255 },
      { id: 4, name: 'Black',  r: 0, g: 0, b: 0 },
      { id: 5, name: 'Grey',   r: 122, g: 122, b: 122 }
    ],
    pixels: [...], 
    filters: [
      { id: 0, type: 'pixelmap', name: 'Blur', matrix: [0,1,1,0,0,1,0,...] },
      { id: 1, type: 'colormap', name: 'Bichromal', colors: [
          { id: 3, name: 'White',  r: 255, g: 255, b:255 },
          { id: 5, name: 'Grey',   r: 122, g: 122, b: 122 }
        ] 
      }
    ] 
  }
```

Ok, so the things I'm (mostly) stuck on are:

1. How to define Models that are shared by more than one Model.  (I'm guessing this may be through "hasManyBy()" and the (future) "hasOneBy()" functions, but I'm not sure any more.

2. How to define the "bitmaps.filters" property, which can contain a variety of other models.

Any advice appreciated!

Thanks.

**Comment by kiaking** [comment]
Thanks for the example! Now I understand.

> 1. How to define Models that are shared by more than one Model. (I'm guessing this may be through "hasManyBy()" and the (future) "hasOneBy()" functions, but I'm not sure anymore.

At first, the Bitmap and Platte should have "HasAndBelongsToMany" relationship which Vuex ORM doesn't support yet. I should implement it. (many to many relations is must-have feature for the ORM anyway).

After implementing many to many relations, you first define the "pivot table" model. And make them associate with the pivot model. 

```js
// Omitting unrelated fields.

class Bitmap extends Model {
  static fields () {
    return {
      id: this.attr(null),
      palattes: this.belongsToMany(Palette, BitmapPalette, 'bitmap_id', 'palette_id')
    }
  }
}

class Palette extends Model {
  static fields () {
    return {
      id: this.attr(null),
      bitmaps: this.belongsToMany(Palette, BitmapPalette, 'palette_id', 'bitmap_id')
    }
  }
}

// Pivot table class. Name could be anything.
class BitmapPalette extends Model {
  static fields () {
    return {
      id: this.attr(null),
      bitmap_id: this.attr(null),
      palatte_id: this.attr(null)
    }
  }
}
```

Next, for Color, ColorMap, and Palette, you want to have "Many To Many Polymorphic Relations" (this is what Laravel calls it) which is also not implemented yet. After implementing morph relation, to share Color which belongs to both ColorMap and Palette, the API should look something like below.

```js
class Palette extends Model {
  static fields () {
    return {
      id: this.attr(null),
      bitmaps: this.belongsToMany(Palette, BitmapPalette, 'palette_id', 'bitmap_id')
      colors: this.morphToMany(Color, Colorable, 'color_id', 'colorable_id', 'colorable_type')
    }
  }
}

class ColorMap extends Model {
  static fields () {
    return {
      id: this.attr(null),
      colors: this.morphToMany(Color, Colorable, 'color_id', 'colorable_id', 'colorable_type')
    }
  }
}

class Color extends Model {
  static fields () {
    return {
      id: this.attr(null),
    }
  }
}

class Colorable extends Model {
  static fields () {
    return {
      color_id: this.attr(null),
      colorable_id: this.attr(null),
      colorable_type: this.attr(null)
    }
  }
}
```


> 2. How to define the "bitmaps.filters" property, which can contain a variety of other models.

This can also be done by Polymorphic Relations once it is implemented.

```js
// Omitting unrelated fields.

class Bitmap extends Model {
  static fields () {
    return {
      id: this.attr(null),
      filters: this.hasMany(Filter, 'bitmap_id')
    }
  }
}

class PixelMap extends Model {
  static fields () {
    return {
      id: this.attr(null)
    }
  }
}

class ColorMap extends Model {
  static fields () {
    return {
      id: this.attr(null),
      colors: this.morphToMany(Color, Colorable, 'color_id', 'colorable_id', 'colorable_type')
    }
  }
}

// Not sure how to define this model...
class Filter extends Model {
  static fields () {
    return {
      id: this.attr(null),
      bitmap_id: this.attr(null),
      mappable_id: this.attr(null),
      mappable_type: this.attr(null),
      map: this.morphTo('mappable_id', 'mappable_type') // <- Get ColorMap or PixelMap
    }
  }
}
```

If you can do this, then it should be able to get the following result.

```js
RESULT: {
  id: 1, 
  name: 'Mona Lisa', 
  palette: { 
    id: 0, 
    name: 'Greyscale', 
    colors: [...],
    pixels: [...], 
    filters: [
      {
        id: 0,
        bitmap_id: 1,
        mappable_id: 10,
        mappable_type: 'PixelMap',
        map: { id: 10 name: 'Blur', matrix: [0,1,1,0,0,1,0,...] }
      },
      {
        id: 1,
        mappable_id: 7,
        mappable_type: 'ColorMap',
        name: 'Bichromal',
        map: {
          id: 7,
          colors: [
            { id: 3, name: 'White',  r: 255, g: 255, b:255 },
            { id: 5, name: 'Grey',   r: 122, g: 122, b: 122 }
          ]
        } 
      }
    ] 
  }
}
```

The object inside `filters` are a bit different from your example, but I think it gets what you need. 

So it has been said, we need to implement those relationships to Vuex ORM to satisfy your needs. I'm willing to do it because these are the feature what ORM must have.

However, because these are complicated stuff, it will take time I guess. And for now, I don't think there is an easy workaround for this.

You need to query the records multiple time and construct the final data by your self.

Let mw implement these features one by one. If you have any questions, please feel free to ask!

**Comment by CookieJon** [comment]
Wow! Thanks for such a detailed explanation. 

Ok, so now I see that it's just a matter of these features not having been implemented yet, I don't feel so stupid not being able to work it out. :-) 

After reading a bit more I get that ORM really is just relational DB. Makes total sense.

Please don't implement anything in your library just on my request (it's so not important!) although I guess as you sat ManyToMany, etc. are features you'd probably want to add anyway.

I think I'll stick with it, use some workarounds for the moment, and check back here later after some time.

Thanks again!

**Comment by kiaking** [comment]
> After reading a bit more I get that ORM really is just relational DB. Makes total sense.

Yeah it's trying to be. Bit different from other ORM since usually ORM generates SQL, instead Vuex ORM manipulates JSON (Vuex Store State) through ORM like api syntax.

> Please don't implement anything in your library just on my request (it's so not important!) although I guess as you sat ManyToMany, etc. are features you'd probably want to add anyway.

Yeah as you mentioned, I'll implement those anyway! Thanks for bringing up though. I had ManyToMany relationship on my ToDo list for long but since no one was asking for it, it kept dropping its priority lol

**Comment by kiaking** [comment]
The first requirement for this issue – Many To Many – relation is now available with Vuex ORM 0.17.0! 🎉  Please [see the doc](https://vuex-orm.gitbooks.io/vuex-orm/relationships/defining-relationships.html#many-to-many) on how to define it.

**Comment by DennisOng** [comment]
Just started using this package for the past few days and just wanted to say thanks for the superb effort @kiaking ! Also went down the route implementing a similar thing manually using normalizr and things was getting messy, this package really simplifies that.

Also wanted to add that "Many To Many Polymorphic Relations" would be super useful for our use-case, but trying to workaround it now using plain many-to-many relations.

**Comment by kiaking** [comment]
@DennisOng Thank you for using Vuex ORM! Yes, I'm working on with morph relation currently : ) It's bit complicated, so it might take time but stay tuned! I wouldn't think it takes too long though.

**Comment by kiaking** [comment]
Polymorphic relations are implemented by #86! It will be available on next release 🚀

**Comment by kiaking** [comment]
Now the Vuex ORM 0.20.0 supports polymorphic relation 🎉 
https://github.com/vuex-orm/vuex-orm/releases/tag/v0.20.0

Thank you all for being patient on this, and thank you all for bringing this up!
